### PR TITLE
Attempt to keep the file extension intact when downloading files

### LINKED ISSUE #3715: Download machinery should attempt to keep the file extension of the downloaded URL intact, if one exists
**Version**

All

**Describe the bug**

Sometimes libraries will get upset when you pass into them a file without the file extension they expect for the type of file they are handling.  It is best to keep the file extension intact to avoid those kinds of issues.

See https://bugzilla.redhat.com/show_bug.cgi?id=2176870

There is a deeper issue there with the `file` / `libmagic` library, but until that is fixed and shipped the best we can do is help it out by keeping the original filenames.

I'm sure it will also help with debugging on occasion.

**To Reproduce**
See https://bugzilla.redhat.com/show_bug.cgi?id=2176870

**Additional context**
BZ https://bugzilla.redhat.com/show_bug.cgi?id=2176870

### PR DESCRIPTION
**Pull request by dralley**
closes #3715

### CODE REVIEW

File: pulpcore/download/base.py [hunk 1]
```diff
@@ -116,7 +116,14 @@ def _ensure_writer_has_open_file(self):
...
         if not self._writer:
-            self._writer = tempfile.NamedTemporaryFile(dir=".", delete=False)
+            filename = self.url.split("/")[-1]
+            # this should be safe even with URL encoding - linux allows any character except NUL or / in a filename
+            # making it urlencoding aware would be nice, but not critical
+            is_legal_filename = filename and len(filename <= 255)
+            # if the url ends with / then we just fall back to providing a random name
+            suffix = "-" + filename if is_legal_filename else None
+            # write the file to the current working directory with a random prefix and the desired suffix
+            self._writer = tempfile.NamedTemporaryFile(dir=".", suffix=suffix, delete=False)
```
Comment (dralley) [inline pulpcore/download/base.py:126]: Am I missing any corner cases?

Comment (ggainey) [reply pulpcore/download/base.py:129]: 255 bytes is max-filename-length in Linux - if you're adding a "-", shouldn't we be checking against 254?

Comment (dralley) [reply pulpcore/download/base.py:129]: Right, actually less than that because of the random prefix

**Review by ggainey — APPROVED** [review]
In the (pathological) case of the filename being > 243 bytes, we *might* still run into the problem of #3715, yes? But this improves our current situation enough to accept, while we hope for a real fix from the upstream library.

File: pulpcore/download/base.py (hunk 1 above)
Comment (dralley) [reply pulpcore/download/base.py:129]: Hmmm what about trailing GET parameters?  Crap, I think that would break it.

Comment (ggainey) [reply pulpcore/download/base.py:129]: Maybe use urllib.parse(), and then operate on path?

Comment (dralley) [reply pulpcore/download/base.py:130]: Done

Comment (ggainey) [reply pulpcore/download/base.py:130]: There we go!

**Review by ggainey — APPROVED** [review]

**Review by decko — APPROVED** [review]

**Comment by dralley** [comment]
@mdellweg / @ipanova Keeping your review requests as it would be good for someone who works on the container plugin to review also as your use cases are different

**Comment by dralley** [comment]
Oh.  Ok

**Comment by ggainey** [comment]
> @mdellweg / @ipanova Keeping your review requests as it would be good for someone who works on the container plugin to review also as your use cases are different

Ugh - we must have hit buttons at the exact same moment.

**Comment by lubosmj** [comment]
@dralley, this should not affect the pulp_container workflows. There are no file extensions taken into consideration in registries.

Does the default limit equal to 255 characters per filename or absolute path? The latter could be a problem because we already have larger directory names for tasks' working directories.

**Comment by ggainey** [comment]
> Does the default limit equal to 255 characters per filename or absolute path? The latter could be a problem because we already have larger directory names for tasks' working directories.

This is the **filename** limit - see https://www.linuxshelltips.com/find-limit-file-name-length-linux/ for some gory details. Note that a "directory" is a file for this purpose  (like everything else in Linux :) )  - so a single directory can't have a name longer than this max, but a directory-path can be...quite long.

Note also that this isn't a **Pulp** limitation, it's a **filesystem** limitation.

**Pull request merged**